File: 백준/Gold/2631. 줄세우기/줄세우기.c
#include<stdio.h>
#include<stdlib.h>

int main(void)
{
	int N, *children=NULL, *LIS=NULL, max=1;

	scanf("%d", &N);
	children=(int *)malloc(N*sizeof(int));
	LIS=(int *)malloc(N*sizeof(int ));
	for(int n=0;n<N;n++)
		scanf("%d", &children[n]);

	for(int i=N-1;i>=0;i--)
	{
		LIS[i]=1;
		for(int j=i+1;j<N;j++)
			if(children[i]<children[j] && LIS[i]<=LIS[j]+1)
				LIS[i]=LIS[j]+1;
		if(max<LIS[i])
			max=LIS[i];
	}

	printf("%d\n", N-max);
	free(children);
	free(LIS);
	return 0;
}

File: 백준/Gold/2631. 줄세우기/줄세우기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		
		int N=Integer.parseInt(br.readLine()), max=1;
		int[] children=new int[N], LIS=new int[N];
		
		for(int n=0;n<N;n++)
			children[n]=Integer.parseInt(br.readLine());
		
		for(int i=N-1;i>=0;i--) {
			LIS[i]=1;
			for(int j=i+1;j<N;j++)
				LIS[i]=children[i]<children[j]&&LIS[i]<=LIS[j]+1?LIS[j]+1:LIS[i];
			max=LIS[i]>max?LIS[i]:max;
		}
		
		System.out.println(N-max);
	}
}